### PR TITLE
fix(_comp_command_offset): Add support for complete -C

### LINKED ISSUE #631: Use `sudo` with `complete -C` commands
## Describe the bug

When a command is completed with `complete -C`, sudo complete is not working.

## To reproduce

```
$ complete -C 'echo' foo
$ sudo foo <tab>
```

Gives warning:

```
-bash: compgen: warning: -C option may not work as you expect
```

## Expected behavior

No warning message thrown, and `sudo foo` completes the same way as `foo`.

## Versions (please complete the following information)

- [linux] Operating system name/distribution and version:
- [5.1] bash version, `echo "$BASH_VERSION"`:
- [2.11] bash-completion version, `(IFS=.; echo "${BASH_COMPLETION_VERSINFO[*]}")`:

## Additional context

`man bash` says about `compgen`:

>   When using the -F or -C options, the various shell variables set by the programmable completion facilities, while available, will not have useful values.

`_command_offset` knows about `-F` but not `-C`:

```
...
if [[ ${cspec#* -F } != "$cspec" ]]; then
...
```

### PR DESCRIPTION
**Pull request by elyscape**
Fixes #631.

### CODE REVIEW

File: test/t/unit/test_unit_command_offset.py [hunk 1]
```diff
@@ -51,6 +56,7 @@ def test_1(self, bash, functions):
             ("cmd3", wordlist),
             ("cmd4", []),
             ("cmd5", ["0"]),
+            ("cmd6", wordlist),
```
Comment (akinomyoga) [inline test/t/unit/test_unit_command_offset.py:59]: Could you separate the test from `test_2`? As in the doc comment, `test_2` is the test template for the completion retry requested by the exit status 124.

Also, the test seems to fail in my environment currently. I wondered how this could get passed [CI test #601](https://github.com/scop/bash-completion/actions/runs/2968137118), but I noticed that actually the test `test_unit_command_offset.py` isn't run in the CI test because it is not included in the [EXTRA_DIST](https://github.com/scop/bash-completion/blob/40aa4dc3c75791f43f850871466039768b3048ad/test/t/Makefile.am#L3)...

**Review by akinomyoga — CHANGES_REQUESTED** [review]
LGTM except for the following minor points.

Also, I think we can add test cases for the line continuation by the trailing `\`.

File: bash_completion [hunk 2]
```diff
@@ -2274,33 +2274,85 @@ _command_offset()
...
+                        # strip trailing newlines manually
+                        suggestion="${suggestion%$'\n'}"
+
+                        COMPREPLY[i]="${COMPREPLY[i]-}${suggestion%\\}"
+
+                        if [[ $suggestion == *\\ ]]; then
+                            COMPREPLY[i]+=$'\n'
+                        else
+                            ((i++))
+                        fi
```
[the remark this answers is not in the document]
Comment (elyscape) [reply bash_completion]: Done.

File: test/t/unit/test_unit_command_offset.py (hunk 1 above)
Comment (elyscape) [reply test/t/unit/test_unit_command_offset.py]: Separated, and tests enhanced to actually work.

**Comment by elyscape** [comment]
I wasn't able to figure out a way to test for completion after the newline is inserted. That being said, it does seem to function as expected, and I expect that it doesn't actually come up that often in practice.

File: test/t/unit/test_unit_command_offset.py [hunk 3]
```diff
@@ -30,9 +31,12 @@ def functions(self, bash):
             'complete -W %s "${COMP_WORDS[0]}"' % quote(join(self.wordlist)),
             'COMPREPLY=(dummy); complete -r "${COMP_WORDS[0]}"',
             "COMPREPLY+=(${#COMPREPLY[@]})",
+            'complete -C ./completer "${COMP_WORDS[0]}"',
         ]
 
         for idx, comp in enumerate(completions, 2):
+            if comp is None:
+                continue
```
Comment (akinomyoga) [inline test/t/unit/test_unit_command_offset.py:39]: This is now unused.

File: test/t/unit/test_unit_command_offset.py [hunk 4]
```diff
@@ -30,9 +31,12 @@ def functions(self, bash):
             'complete -W %s "${COMP_WORDS[0]}"' % quote(join(self.wordlist)),
             'COMPREPLY=(dummy); complete -r "${COMP_WORDS[0]}"',
             "COMPREPLY+=(${#COMPREPLY[@]})",
+            'complete -C ./completer "${COMP_WORDS[0]}"',
```
Comment (akinomyoga) [inline test/t/unit/test_unit_command_offset.py:34]: This array is also for 124. I think the following line should be added at the end of `functions(self,bash)` instead:

```bash
        assert_bash_exec(bash, "complete -C ./completer cmd6")
```

File: test/fixtures/_command_offset/completer [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        echo two
+        echo long
+        ;;
+    li*)
+        echo line\\
+        echo two
+        ;;
+    lo*)
+        echo long
+        ;;
```
Comment (akinomyoga) [inline test/fixtures/_command_offset/completer:21]: I'd like to add the corner case where the last line is terminated by `\`. What do you think?

```suggestion
        ;;
    cont1*)
        echo cont10
        echo cont11\\
        ;;
```

File: test/t/unit/test_unit_command_offset.py [hunk 6]
```diff
@@ -72,3 +76,17 @@ def test_2(self, bash, functions, cmd, expected_completion):
...
         """
         assert assert_complete(bash, "meta %s " % cmd) == expected_completion
+
+    @pytest.mark.parametrize(
+        "cmd,expected_completion",
+        [
+            ("cmd6 ", wordlist),
+            ("cmd6 l", [r"line\^Jtwo", "long"]),
+            ("cmd6 lo", ["ng"]),
+            ("cmd6 line", [r"\^Jtwo"]),
```
Comment (akinomyoga) [inline test/t/unit/test_unit_command_offset.py:86]:
```suggestion
            ("cmd6 line", [r"\^Jtwo"]),
            ("cmd6 cont1", ["cont10", "cont11\\"]),
```

**Review by akinomyoga — CHANGES_REQUESTED** [review]
Thank you for updating! These are the remaining points.

File: test/fixtures/_command_offset/completer (hunk 5 above)
Comment (eliyoung-ethos) [reply test/fixtures/_command_offset/completer:21]: Assuming that's how bash handles it, sure.

**Comment by eliyoung-ethos** [comment]
Changes applied.

File: test/fixtures/_command_offset/completer (hunk 5 above)
Comment (elyscape) [reply test/fixtures/_command_offset/completer:25]: Assuming that's how bash handles it, sure.

**Comment by elyscape** [comment]
Requested changes applied! Let me know if there's anything else I should change.

File: test/t/unit/test_unit_command_offset.py (hunk 4 above)
Comment (akinomyoga) [reply test/t/unit/test_unit_command_offset.py]: Thank you!

You have put the above line in `test_3` instead of in `functions`. I'm interested in if there is a reason that this line is put directly in `test_3`. If we put it in `test_3`, it will be executed every time for each `(cmd,expected_result)` pair. But I think this common setup can be executed just once, which is one of the reasons that one defines completion settings with `@pytest.fixture`.

Comment (elyscape) [reply test/t/unit/test_unit_command_offset.py]: Whoops! I got confused and put it in the wrong place. Fixing now.

**Comment by akinomyoga** [comment]
Thank you! Now I think everything is OK. Let's wait for @scop's check and decision!

**Review by akinomyoga — APPROVED** [review]

**Comment by elyscape** [comment]
I went ahead and rebased onto master to pick up the changes in #794.

**Comment by elyscape** [comment]
Rebased onto master to handle changes from #797.

**Comment by elyscape** [comment]
Just checking in case this fell through the cracks.

File: bash_completion [hunk 7]
```diff
@@ -2307,33 +2307,83 @@ _comp_command_offset()
...
-                    continue
+                    $reset_monitor
+                    $reset_lastpipe
+
+                    local suggestion
+                    local i=0
+                    COMPREPLY=()
+                    for suggestion in "${suggestions[@]}"; do
+                        # `mapfile -t` was added in bash 4.4, so we have to
+                        # strip trailing newlines manually
```
Comment (scop) [inline bash_completion:2378]: Are we sure about this? We've used `mapfile -t` in the `cvs` completion since 2011 (e35eabb791733996078aae9161fc01ab5f94a9e3), I think 5 years before bash 4.4 was released.

File: bash_completion [hunk 8]
```diff
@@ -2307,33 +2307,83 @@ _comp_command_offset()
...
+                    # completer commands are always single-quoted
+                    if ! _comp_dequote "'$completer"; then
+                        _minimal
+                        break
+                    fi
+                    completer=${ret[0]}
+
+                    local -a suggestions
+
+                    local reset_monitor=$(shopt -po monitor) reset_lastpipe=$(shopt -p lastpipe)
```
Comment (scop) [inline bash_completion:2355]: Do we need protection against unusual `$IFS` here (at least for eventually invoking the resets)? Cf. `git grep -F ' reset'`.

**Review by scop — CHANGES_REQUESTED** [review]
Sorry for sitting on this for so long. Added a few comments + looks like we need another rebase.

**Comment by elyscape** [comment]
I'll take a look this weekend.

File: bash_completion (hunk 7 above)
Comment (elyscape) [reply bash_completion:2378]: The `NEWS` file indicates that [it was added in 4.4][news-4.4], but doing some digging, that's it looks like that notice was misleading. Looking at the source code, it looks like the flag did exist starting with bash 4.0, but originally [only removed trailing newlines][mapfile-4.0]; bash 4.4 added the `-d` flag and updated `-t` to remove the specified delimiter when both flags are used. I'll update this accordingly.

[news-4.4]: https://github.com/bminor/bash/blob/bash-4.4/NEWS#L17-L19
[mapfile-4.0]: https://github.com/bminor/bash/blame/bash-4.0/builtins/mapfile.def#L37

File: bash_completion (hunk 8 above)
Comment (elyscape) [reply bash_completion:2355]: Good point. Will fix.

**Comment by elyscape** [comment]
Rebased again and addressed comments. We should be good to go now.

**Comment by elyscape** [comment]
@scop Just checking in on this.

**Comment by akinomyoga** [comment]
rebased.

**Review by akinomyoga — APPROVED** [review]

**Comment by elyscape** [comment]
@scop Pinging on this.

**Review by scop — APPROVED** [review]
LGTM. Thanks for your patience, this took way too long, my bad.
